src/python/pants/backend/scala/bsp/rules.py

```diff
@@ -138,10 +172,98 @@ async def bsp_scalac_options_request(request: ScalacOptionsParams) -> ScalacOpti
     return ScalacOptionsResult(items=tuple(result.item for result in results))
 
 
+# -----------------------------------------------------------------------------------------------
+# Compile Request
+# -----------------------------------------------------------------------------------------------
+
+
+@dataclass(frozen=True)
+class ScalaBSPCompileFieldSet(BSPCompileFieldSet):
+    required_fields = (ScalaSourceField,)
+    source: ScalaSourceField
+
+
+@rule
+async def bsp_scala_compile_request(
+    request: ScalaBSPCompileFieldSet,
+    union_membership: UnionMembership,
+    unzip: UnzipBinary,
+    bash: BashBinary,
+) -> BSPCompileResult:
+    coarsened_targets = await Get(CoarsenedTargets, Addresses([request.source.address]))
+
+    # NB: Each root can have an independent resolve, because there is no inherent relation
+    # between them other than that they were on the commandline together.
+    resolves = await MultiGet(
+        Get(CoursierResolveKey, CoarsenedTargets([t])) for t in coarsened_targets
+    )
+
+    results = await MultiGet(
+        Get(
+            FallibleClasspathEntry,
+            ClasspathEntryRequest,
+            ClasspathEntryRequest.for_targets(union_membership, component=target, resolve=resolve),
+        )
+        for target, resolve in zip(coarsened_targets, resolves)
+    )
+    _logger.info(f"results = {results}")
+
+    exit_code = next((result.exit_code for result in results if result.exit_code != 0), 0)
+    _logger.info(f"exit code = {exit_code}")
+    output_digest = EMPTY_DIGEST
+    if exit_code == 0:
+        digests: list[Digest] = []
+        filenames: list[str] = []
+        for result in results:
+            if result.output:
+                digests.append(result.output.digest)
+                filenames.extend(result.output.filenames)
+        jars_digest = await Get(Digest, MergeDigests(digests))
+        input_digest = await Get(Digest, CreateDigest([Directory("__classpath__")]))
+        script_digest = await Get(
```

Comment:
Is there a better way to expand the classpath jars of first-party sources to classfiles?